BoundedSumKnownN.tex

```diff
@@ -0,0 +1,351 @@
+    input_metric = SymmetricDistance()
+    output_metric = AbsoluteDistance(T)
+    
+    if L*n < get_min_value(T) or U*n > get_max_value(T): 
+        raise Exception('Invalid parameters') |\label{line:max}|
+    
+    def Relation(d_in: u32, d_out: u32) -> bool: |\label{line:rel}|
+        return d_out >= d_in*(U-L)/2
+    
+    def function(data: Vec(T)) -> T: |\label{line:fn}|
+        return data.iter().sum() |\label{line:sum}|
+    return Transformation(input_domain, output_domain, function, input_metric, output_metric, stability_relation)
+\end{lstlisting}
+
+\iffalse
+\subsubsection*{Conditions as specified in the pseudocode}
+\begin{itemize}
+    \item \underline{Input domain:} sized domain of all vectors of elements in \texttt{IntervalDomain(L, U)}, where \texttt{L} and \texttt{U} are of type \texttt{T} and stand for \textit{lower bound} and \textit{upper bound}, respectively.\footnote{The compiler infers \texttt{float} from the context, which is why it is omitted.} 
+    
+    \item \underline{Output domain:} domain of all elements of type \texttt{T}.
+    
+    \item \underline{Function:} given $v$ of type \texttt{Vec(T)}, it returns an element of type \texttt{T} which corresponds to the sum of all the vector entries $v_i$ of $v$.
+    
+    \item \underline{Input metric:} symmetric distance.\footnote{As of June 24, SubstituteDistance is no longer a metric part of the OpenDP library.}
+    
+    \item \underline{Output metric:} absolute distance.
+    
+    \item \underline{Stability relation:} for $\din, \dout$ of type \texttt{u32}, the relation returns $\True$ if and only if $\dout \geq \din \cdot (\U-\texttt{L})/2$.
+\end{itemize}
+\fi
+
+%More concisely, the stability relation can be stated as:%\footnote{See page 14 of \textit{A Programming Framework for OpenDP} for more examples. Note that in the Rust implementation the metrics $d_{\X}$ and $d_{\Y}$ are \textit{not} inputs to the relation, and they are instead fixed as attributes of the transformation.}
+%\begin{equation}\label{eq:relation}
+    %\textrm{R}(\din, \dout) = 
+    %\begin{cases} 
+      %\True & \textrm{if } \dout \geq \din \cdot (\U - \texttt{L})/2 \textrm{ for } \din, \dout \textrm{ of type } \texttt{u32} \\
+      %\False & \textrm{otherwise}
+   %\end{cases}
+%\end{equation}
+%\silvia{Maybe take out the equation form; repetitive.}
+
+\section{Proof}
+\subsection{Symmetric Distance}
+\begin{theorem}
+    For every setting of the input parameters \texttt{(L, U, n)} to \texttt{MakeBoundedSumN}, the transformation returned by \texttt{MakeBoundedSumN} has the following properties:
+    \begin{enumerate}
+        \item \textup{(Appropriate output domain).} For every element $v$ in \texttt{input\_domain}, $\function(v)$ is in \texttt{output\_domain}. % Prof. Vadhan said on 29/6 to leave it as this for now, but maybe in the future we add to the theorem statement what exactly the input domain and output domain are (although of course we already say this in the proof)
```

Comment:
Do we not have to add that "otherwise the program will raise an Exception"?

Reply:
for line 186

Reply:
I don't think so, because that is the point of having theorems that prove properties about the _returned constructor_ by the function: that is, the theorem is not about the program itself, but rather about the constructor that the program outputs. If the program is correct, then the constructor that it returns _must_ return an `output_domain` which is correct. Any exceptions for failures need to happen during the execution of the program, in which case a constructor will not be returned, which is why the theorem is not concerned with this case. (I am assuming that if the program raises an Exception, it stops running and does not return any constructor; is that what you are thinking too?)